proof/invariant-abstract/AInvs.thy

```diff
@@ -9,7 +9,7 @@ The main theorem
 *)
 
 theory AInvs
-imports ArchAInvsPre
+imports ArchDetSchedSchedule_AI
```

Comment:
Here are a couple, perhaps left-field, thoughts on where to put invariants versus where to put specs and proofs.

I was thinking about how much effort would be required to get rid of the Haskell. (It might be fun to discuss this further.) Wanting to assert properties about the `kernel_state` (that is, the state at the Refine level) within functions means that we would want some way of stating the invariants at that level before we write the functions at that level (unless we resorted to some other kind of trickery, which I would be against). I know that there was some discussion of having a separate session for all the `invs'` lemmas in Refine, and having the `corres` rules elsewhere. So something like invariants, spec, Hoare triples, corres rules might be ultimately the best approach.

Considering that we're experimenting with heap projections and how best to handle them, it might be good to move this work as high as possible so that we can use these definitions in the abstract spec. Or is there some reason this would be a bad idea.

And while preparing my talk a few weeks ago, I read some people, especially mathematicians, who were quite interested in the formalisation of particular (mostly mathematical) concepts, but less interested in the proofs: they were willing to accept the correctness of the theorem provers, but were more hesitant to accept that the formalisations captured the mathematical concepts they were interested in. So they wanted to see the definitions separate from the proofs. Would it be nicer for us to be more strict in separating out the invariants and other definitions from the proofs themselves?

Reply:
(Well, in order to define invariants over the state, we need to define the state. So maybe it's more like definition of the state, invariants, functions in the kernel, Hoare triples, corres rules, or something. But the point remains that we might benefit from moving the definition of the invariants up higher, or as high as possible.

And I can't think up an example right now where I've thought we'd benefit from having the invariants or other things in AInvs in the abstract spec. But we'd definitely like some of the things from, say, `DetSchedInvs_AI.thy` further up in AInvs)

Reply:
I don't think we need to keep all definitions apart from all proofs, but it is usually good to at least group definitions together. It often makes sense to have basic properties about definitions in the same file, both for reading and developing, but it can become too unwieldy. It requires some balance.

In any case, I do like AInvs being the toplevel theory again, and we should indeed update the docs (happy for a separate PR on that).